agent/csi/plugin/plugin.go

```diff
@@ -263,11 +263,8 @@ func (np *nodePlugin) NodeUnstageVolume(ctx context.Context, req *api.VolumeAssi
 
 	// we must unpublish before we unstage. verify here that the volume is not
 	// published.
-	if v, ok := np.volumeMap[req.ID]; ok {
-		if v.isPublished {
-			return status.Errorf(codes.FailedPrecondition, "Volume %s is not unpublished", req.ID)
-		}
-		return nil
+	if v, ok := np.volumeMap[req.ID]; ok && v.isPublished {
+		return status.Errorf(codes.FailedPrecondition, "Volume %s is not unpublished", req.ID)
 	}
 
```

Comment:
I'm not super-familiar with this code, but should this actually proceed if the volume was not found in `np.volumeMap` ?

i.e., the volume should be present if it was successfully staged. If a failure happened during staging, it wouldn't be added. Looking at `NodeStageVolume` further up; https://github.com/moby/swarmkit/blob/e8ecf83ee08e14a05e28992dc304576079d403c7/agent/csi/plugin/plugin.go#L231-L239


Perhaps this should be something like;


```go
if v, ok := np.volumeMap[req.ID]; !ok || v.isPublished {
	// volume not found or is still published
	return status.Errorf(codes.FailedPrecondition, "Volume %s is not unpublished", req.ID)
}
```

Or if it would be useful to have distinct errors for each situation

```go
v, ok := np.volumeMap[req.ID
if !ok {
	return status.Errorf(codes.FailedPrecondition, "Volume %s not found", req.ID)
}
if v.isPublished {
	return status.Errorf(codes.FailedPrecondition, "Volume %s is not unpublished", req.ID)
}
```

Reply:
Indeed the assumptions around state for unstaging would always hold true if `np.volumeMap` was made persistent: 
https://github.com/moby/swarmkit/blob/ea1a7cec35cb05dce3cff2de3f4ee28cbca6ca79/agent/csi/plugin/plugin.go#L60-L63

I recall during testing it is possible that a volume had been staged before the node daemon restarted and `np.volumeMap` was empty. In the method `NodeUnpublishVolume` it always unpublishes the volume irrespective of `np.VolumeMap`.